File: java/StringtoInteger.java
/*
 * Filename: StringtoInteger.java
 * Author: Jun Xu
 * Email: junx@bu.edu
 * Created Time: Wed 04 Mar 2015 07:28:00 PM EST
 */

public class Solution {
	public int atoi(String str) {
		int i = 0;

		while (i < str.length() && (int) str.charAt(i) == 32) {
			i++;
		}

		if (i == str.length()) {
			return 0;
		}

		int sign = 1;
		if ((int) str.charAt(i) == '+') {
			i++;
		} else if ((int) str.charAt(i) == '-') {
			i++;
			sign = -1;
		}

		int ret = 0;
		int digit = (int) str.charAt(i) - '0';
		if (digit < 0 || digit > 9) {
			return 0;
		}

		for (; i < str.length(); i++) {
			digit = (int) str.charAt(i) - '0';

			if (digit < 0 || digit > 9) {
				break;
			}

			if (sign == 1) {
				if (ret > Integer.MAX_VALUE / 10 ||
						(ret == Integer.MAX_VALUE / 10 && digit > Integer.MAX_VALUE % 10)) {
					return Integer.MAX_VALUE;
				} else {
					ret = ret * 10 + digit;
				}
			} else {
				if (ret * sign < Integer.MIN_VALUE / 10 ||
						(ret * sign == Integer.MIN_VALUE / 10 && digit * sign < Integer.MIN_VALUE % 10)) {
					return Integer.MIN_VALUE;
				} else {
					ret = ret * 10 + digit;
				}
			}
		}

		return ret * sign;
	}
}

File: cpp/stringToInteger.cpp
/*
 * Filename: stringToInteger.cpp
 * Author: Jun Xu
 * Email: junx@cs.bu.edu
 * Created Time: Thu 21 Jan 2016 08:55:42 PM EST
 * Description: 
 */

#include <climits>
#include <iostream>

using namespace std;

class Solution
{
	public:
		int myAtoi(string str)
		{
			bool pos = true;
			int i = 0;
			int digit;
			int res = 0;

			while (i < str.length() && str[i] == ' ')
			{
				i++;
			}

			if (i == str.length())
			{
				return 0;
			}
			else if (str[i] == '+')
			{
				i++;
			}
			else if (str[i] == '-')
			{
				i++;
				pos = false;
			}

			while (i < str.length())
			{
				digit = str[i] - '0';
				if (digit < 0 || digit > 9)
				{
					break;
				}

				if (pos && (res > INT_MAX / 10 || (res == INT_MAX / 10 && digit > INT_MAX % 10)))
				{
					return INT_MAX;
				}
				else if (!pos && (-res < INT_MIN / 10 || (-res == INT_MIN / 10 && -digit < INT_MIN % 10)))
				{
					return INT_MIN;
				}

				i++;
				res = res * 10 + digit;

				cout << res << endl;
			}

			return pos ? res : -res;
		}
};
